ktor-client/ktor-client-cio/jvm/src/io/ktor/client/engine/cio/Proxy.jvm.kt

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2014-2025 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.engine.cio
+
+import io.ktor.client.engine.ProxyConfig
+import java.net.InetSocketAddress
+import java.net.Proxy
+
+internal actual fun discoverHttpProxy(): ProxyConfig? {
+    var proxy: Proxy? = null
+
+    try {
+        // For Android
+        val host = System.getProperty("http.proxyHost")
+        val port = System.getProperty("http.proxyPort")?.toIntOrNull() ?: 0
```

Comment:
I guess we should rely on `ProxySelector.getDefault()` instead. There is more sophisticated logic to determine proxy host and port (see `sun.net.spi.DefaultProxySelector`).
This would require requesting a new proxy for each URL, but it doesn't seem a problem here.

Reply:
At this point, we don't have the URL to choose the proxy for.

Reply:
> At this point, we don't have the URL to choose the proxy for.

Yes, but we could turn the field `proxy` into a lambda taking an `Url` as a parameter.

> Can you explain why the solution with the ProxySelector is better than the current one?

I see two reasons.
The first one is that the current logic is not enough. Proxy could also be configured via `proxyHost` and `socksProxyHost` system properties. And `ProxySelector` would honor non-proxy hosts configured in the system.
The second reason is that using `ProxySelector` would make CIO behavior consistent to OkHttp and Java engines as they also rely on `ProxySelector`.
As a bonus it would solve [this issue](https://youtrack.jetbrains.com/issue/KTOR-3727/Add-a-way-to-configure-a-proxy-for-each-HTTP-request) (for JVM only).